test/cpp/end2end/time_change_test.cc

```diff
@@ -0,0 +1,364 @@
+      channel->WaitForConnected(grpc_timeout_milliseconds_to_deadline(5000)));
+  auto stub = CreateStub();
+
+  auto stream = stub->BidiStream(&context);
+
+  request.set_message("Hello");
+  EXPECT_TRUE(stream->Write(request));
+  EXPECT_TRUE(stream->Read(&response));
+
+  // time jumps back 20.123 seconds.
+  set_now_offset(-20123);
+
+  request.set_message("World");
+  EXPECT_TRUE(stream->Write(request));
+  EXPECT_TRUE(stream->WritesDone());
+  EXPECT_TRUE(stream->Read(&response));
+
+  auto status = stream->Finish();
+  EXPECT_TRUE(status.ok());
+}
+
+// Wall-clock time jumps forward on client before connection to server is up
+TEST_F(TimeChangeTest, TimeJumpForwardBeforeServerConnect) {
+  EchoRequest request;
+  EchoResponse response;
+  ClientContext context;
+  context.set_deadline(grpc_timeout_milliseconds_to_deadline(5000));
+  context.AddMetadata(kServerResponseStreamsToSend, "2");
+
+  auto channel = GetChannel();
+  GPR_ASSERT(channel);
+
+  // time jumps forward by 5.678 seconds
+  set_now_offset(5678);
+
+  auto ret =
+      channel->WaitForConnected(grpc_timeout_milliseconds_to_deadline(5000));
+  // We use monotonic clock for pthread_cond_timedwait() deadline on linux, and
+  // realtime clock on other platforms - see gpr_cv_wait() in sync_posix.cc.
+  // So changes in system clock affect deadlines on non-linux platforms
+#ifdef GPR_LINUX
+  EXPECT_TRUE(ret);
+  auto stub = CreateStub();
+  auto stream = stub->BidiStream(&context);
+
+  request.set_message("Hello");
+  EXPECT_TRUE(stream->Write(request));
+  EXPECT_TRUE(stream->Read(&response));
+  request.set_message("World");
+  EXPECT_TRUE(stream->Write(request));
+  EXPECT_TRUE(stream->WritesDone());
+  EXPECT_TRUE(stream->Read(&response));
+
+  auto status = stream->Finish();
+  EXPECT_TRUE(status.ok());
+#else
+  EXPECT_FALSE(ret);
+#endif
+}
+
+// Wall-clock time jumps back on client before connection to server is up
+TEST_F(TimeChangeTest, TimeJumpBackBeforeServerConnect) {
+  EchoRequest request;
+  EchoResponse response;
+  ClientContext context;
+  context.set_deadline(grpc_timeout_milliseconds_to_deadline(5000));
+  context.AddMetadata(kServerResponseStreamsToSend, "2");
+
+  auto channel = GetChannel();
+  GPR_ASSERT(channel);
+
+  // time jumps back by 5.678 seconds
+  set_now_offset(-5678);
+
+  EXPECT_TRUE(
+      channel->WaitForConnected(grpc_timeout_milliseconds_to_deadline(5000)));
+  auto stub = CreateStub();
+  auto stream = stub->BidiStream(&context);
+
+  request.set_message("Hello");
+  EXPECT_TRUE(stream->Write(request));
+  EXPECT_TRUE(stream->Read(&response));
+  request.set_message("World");
+  EXPECT_TRUE(stream->Write(request));
+  EXPECT_TRUE(stream->WritesDone());
+  EXPECT_TRUE(stream->Read(&response));
+
+  auto status = stream->Finish();
+  EXPECT_TRUE(status.ok());
+}
+
+}  // namespace
+
+}  // namespace testing
+}  // namespace grpc
+
+int main(int argc, char** argv) {
+  std::string me = argv[0];
+  auto lslash = me.rfind('/');
```

Comment:
I kind of feel `lslash` should be `rslash` here but I might not understood this correctly :)

Reply:
A comment here may be helpful to better understand the if-else block below

Reply:
I think `lslash` = 'last slash'. I got that piece of code from `test/cpp/end2end/server_crash_test.cc`. I have added a comment to explain what's going on here.

Reply:
Ah ok. I thought it was "left slash"